### DISCUSSION TITLE
How to get correct highlights in .svelte files

### DISCUSSION THREAD

**Discussion by gipo355** [HiPhish/rainbow-delimiters.nvim#97] [Q&A]
Hello! thanks for this plugin! 

I was wondering if it was possible to get html inside svelte files to highlight tags differently.

this is the current implementation, all the tags have equal colors
![image](https://github.com/HiPhish/rainbow-delimiters.nvim/assets/91525512/fc76029a-48e1-42c9-a13d-c6715f7244ab)

**Comment by HiPhish** [comment]
It is probably possible, but I don't know Svelte. How do you write files for Svelte? Are they a special file type like Vue? If you could post a sample Svelte file I could look into it.

**Comment by gipo355** [comment]
Of course! thanks for the quick response,

you can find some source files on github @ https://github.com/sveltejs/kit/tree/main/packages/create-svelte/templates/default 

(e.g. https://github.com/sveltejs/kit/blob/main/packages/create-svelte/templates/default/src/routes/sverdle/%2Bpage.svelte)

which is what gets installed when running `npm create svelte@latest my-app` (https://kit.svelte.dev/docs/creating-a-project)

.svelte files include javascript/typescript, HTML, scss/css/postcss in it.

There is its own treesitter which highlights different things but I can't get rainbows in it.

**Comment by gipo355** [comment]
Adding another example page in case something smaller is needed: https://gist.github.com/gipo355/8459cd522df8fa1be82bd23644f06c78

**Comment by HiPhish** [comment]
Is this a reasonable Svelte test file?

```
<script lang="ts">
	import { baz } from 'foo/bar';

	baz({a: {b: {c: 'd'}}});
</script>

<style>
	p {
		font-size: 2em;
	}
</style>

<svelte:head>
	<title>Test page</title>
	<meta name="description" content="A test page" />
</svelte:head>

<h1>A test page for Svelte</h1>

<form action="?/herp/derp" method="post">
	<p>This is a paragraph</p>
	<hr/>
	<ul>
		{#each ["foo", "bar", "baz"] as x}
		<li class="some-class">{ x }</li>
		{/each}
	</ul>
	{# if True}
	<p>Some text </p>
	{/if}
</form>
```

The parser is giving me several problems:

- The CSS inside the `style` is not nested correctly
- The array inside the `each` has errors
- Nested `if`, `elseif` is a real mess because each `elseif` raises the nesting level by one

![Screenshot_20240211_204936](https://github.com/HiPhish/rainbow-delimiters.nvim/assets/4954650/f038f55d-0660-4bb3-bd64-cd672b0affbb)

I don't think this is really production ready and I would rather have someone who knows Svelte help out. Here is the query I got so far: [rainbow-delimiters.txt](https://github.com/HiPhish/rainbow-delimiters.nvim/files/14233808/rainbow-delimiters.txt)

**Reply by gipo355** [reply]
yes, i formatted it and added some nestings


```ts
<script lang="ts">
  import { baz } from 'foo/bar';

  baz({
    a: {
      b: { c: 'd' },
    },
  });

  const x = {
    a: {
      b: {
        c: 'd',
      },
    },
  };
</script>

<svelte:head>
  <title>Test page</title>
  <meta name="description" content="A test page" />
</svelte:head>

<h1>A test page for Svelte</h1>

<div class="1">
  <div class="2">
    <div class="3">
      <div class="4">
        <p class="5"></p>
      </div>
    </div>
  </div>
</div>

<form action="?/herp/derp" method="post">
  <p>This is a paragraph</p>
  <hr />
  <ul>
    {#each ['foo', 'bar', 'baz'] as x}
      <li class="some-class">{x}</li>
    {/each}
  </ul>
  {#if true}
    <p>Some text</p>
  {/if}
</form>

<style lang="scss">
  p {
    font-size: 2em;

    &.some-class {
      color: red;
    }

    h2 {
      color: blue;
    }
  }
</style>
```

![image](https://github.com/HiPhish/rainbow-delimiters.nvim/assets/91525512/7d47f44f-4a2a-4e88-86d3-33229b1bdde4)
![image](https://github.com/HiPhish/rainbow-delimiters.nvim/assets/91525512/d942d148-66e9-4bfe-b513-f11ae6147991)

**Reply by gipo355** [reply]
`:TSInstall svelte` installs tree sitter,

lsps active are
![image](https://github.com/HiPhish/rainbow-delimiters.nvim/assets/91525512/bae26858-5831-4ecc-9a53-ca56c8e2c760)
